src/picai_prep/dcm2mha.py

```diff
@@ -26,372 +29,389 @@
-            image = self.scan_postprocess_func(image)
+class Dicom2MHACase(Case):
+    settings: Dicom2MHASettings = Dicom2MHASettings({}, {})
 
-        try:
-            atomic_image_write(image=image, path=destination, mkdir=True)
-        except Exception as e:
-            item['targets'][target] = f'Error during writing: {e}'
-            self.item_log(item, 'write error')
-            return False
-        return True
+    def __init__(self, input_dir: Path, patient_id: str, study_id: str, paths: List[PathLike]):
+        self.patient_id = patient_id
+        self.study_id = study_id
+        self.series: List[Series] = []
+        self.log = [f'Importing {plural(len(paths), "serie")}']
 
-    def _convert(self):
-        convert_count = 0
-        for item in self.valid_items:
-            for _, status in item['targets'].items():
-                convert_count += 1 if status is None else 0
-
-        self.info(f"Converting {convert_count} DICOM series to MHA.")
-
-        # convert all items with valid targets to mha
-        success_count, skip_count = 0, 0
-        targets = []
-        for item in self.valid_items:
-            for target, status in item['targets'].items():
-                if status is None:  # therefore selected
-                    destination = os.path.join(self.output_dir, item['patient_id'], target + '.mha')
-                    if os.path.exists(destination):
-                        item['targets'][target] = 'Skipped, target output path already exists'
-                        self.item_log(item, 'skipped')
-                        skip_count += 1
-                    else:
-                        targets.append((item, target, destination))
+        fullpaths = set()
+        for path in paths:
+            full_path = input_dir / path
+            serie = Series(full_path, patient_id, study_id)
+            try:
+                if not full_path.exists():
+                    raise ArchiveItemPathNotFoundError(path)
+                elif not full_path.is_dir():
+                    raise NotADirectoryError(path)
+                elif path in fullpaths:
+                    raise FileExistsError(path)
+            except Exception as e:
+                serie.error = e
+                logging.error(str(e))
+            else:
+                fullpaths.add(full_path)
+            finally:
+                self.log.append(f'\t+ ({len(self.series)}) {full_path}')
+                self.series.append(serie)
+        if not all([serie.is_valid for serie in self.series]):
+            self.invalidate()
+
+    @property
+    def valid_series(self):
+        return [item for item in self.series if item.is_valid]
+
+    def invalidate(self):
+        for serie in self.valid_series:
+            serie.error = SeriesException('Invalidated due to critical error in sibling')
+
+    def write_log(self, msg: str):
+        self.log.append(msg)
+
+    def compile_log(self):
```

Comment:
I think this is very nice, but does not outweigh the downsides of error logging: errors should be logged immediately, because who knows what happens afterwards

Reply:
I will explain the reasoning behind the current logging system (so this is a reply to all comment regarding logging)

All exceptions I explicitly raise have varying degrees are severity. Due to multiprocessing, logging "as it happens" is entirely unreadable. We would need to write a whole lot more per logging line to make any sense of it. The current idea is to provide a full and extremely readable log as a case completes. ([example log here](https://pastebin.com/8T8Ximdj))

My `except` are all catching highly specific exceptions that I expect. The user is notified after a case is resolved and may review such a case while the program continues. I may add a summary at the bottom for all errors with any severity for even easier viewing. Any errors I didnt expect are logged immediately with `logging.error` (and will also still show up when the case is complete).

This way, we get best of both worlds: very readable log, with urgent exceptions logged in real time. The line between severity is always blurry and is a case-by-case basis, but I like to think the benefits of the current logs outweigh to drawbacks of non-instant logging of each and every action.

Reply:
Alright, sounds convincing. Thanks for the elaboration!

Reply:
It's also a ton less complicated than the earlier "History" garbage.